### PR TITLE
MTLCompareFunction requires iOS 9.0

### PR DESCRIPTION
**Pull request by Michael-Lfx**
Fixes #issue
PR checklist:
- [x] `make` succeeds (on *nix)
- [ ] `make reftests` succeeds
- [x] tested examples with the following backends:
   - [x]metal
- [x] `rustfmt` run on changed code

### CODE REVIEW

File: src/backend/metal/src/device.rs [hunk 1]
```diff
@@ -1575,7 +1575,10 @@ impl hal::Device<Backend> for Device {
         }
 
         if let Some(fun) = info.comparison {
-            descriptor.set_compare_function(conv::map_compare_function(fun));
+            if (caps.os_is_mac && caps.has_version_at_least(10, 11)) ||
+                (!caps.os_is_mac && caps.has_version_at_least(9, 0)) {
+                descriptor.set_compare_function(conv::map_compare_function(fun));
```
Comment (kvark) [inline src/backend/metal/src/device.rs]: We should return an error otherwise. If the user requests a comparison function, we can't just ignore it.

Comment (Michael-Lfx) [reply src/backend/metal/src/device.rs]: But there is not suitable error of `AllocationError` for the case.

**Comment by kvark** [comment]
We can reuse an existing code and error! the cause, or add a new code aka generic Unsupported.

File: src/backend/metal/src/device.rs (hunk 1 above)
Comment (kvark) [reply src/backend/metal/src/device.rs]: If we can't return an error, we should panic. It's not like the user is going to get anything meaningful anyway if they ask for a comparison sampler but don't get it.

Comment (Michael-Lfx) [reply src/backend/metal/src/device.rs]: Should we add a bool field in `PrivateCapabilities` to indicate that whether `set_compare_function` works on the current platform?

**Comment by kvark** [comment]
Yes, sure!
Note that this issue has been discussed in MoltenVK repo with the conclusion that we still support comparison samplers if they are immutable (provided at the descriptor set layout creation), so your application can still use them. (Link needed)

**Comment by kvark** [comment]
@Michael-Lfx 
> Should we add a bool field in PrivateCapabilities to indicate that whether set_compare_function works on the current platform?

Yes, we need that.
Important to note that we'd still allow the samplers to be created, but disallow them being used in immutable sampler bindings (in `create_xxx_descriptor_sets`).
And as I mentioned, we need to panic right away if this isn't followed by the user.

**Comment by kvark** [comment]
Closed by #2885

**Pull request closed without merging**